docs/copilot/guides/large-complex-codebases-guide.md

```diff
@@ -0,0 +1,408 @@
+---
+ContentId: TODO
+DateApproved: TODO
+MetaDescription: Learn how to set up large complex codebases for AI pair programming.
+MetaSocialImage: ../images/shared/github-copilot-social.png
+---
+# Set up large complex codebases for AI pair programming
+
+This guide shows you how to scale context engineering for large, complex codebases with thousands of files, multiple teams, and intricate architectural dependencies. Building on the [context engineering guide](/docs/copilot/guides/context-engineering-guide.md), this covers advanced strategies for managing AI context in brownfield codebases where traditional "vibe coding" approaches break down.
+
+## Key challenges in large codebases
+
+- **Context overload**: Too much information dilutes AI focus
+- **Team coordination**: Different teams need different AI configurations
+- **Legacy constraints**: Mixed old and new patterns confuse AI
+- **Domain knowledge**: Business logic scattered across the codebase
+
+<!-- TODO: Add step-by-step assessment checklist to identify which challenges apply to your codebase -->
+
+## Hierarchical instruction architecture
+
+### Multi-level instruction files
+
+Unlike single-project setups, large codebases benefit from a hierarchical context structure using VS Code's [custom instructions](/docs/copilot/customization/custom-instructions.md):
+
+```
+.github/
+  copilot-instructions.md          # Repository-wide context (auto-applied)
+  instructions/
+    frontend-dev.instructions.md   # Team-specific context
+    backend-dev.instructions.md
+    platform-dev.instructions.md
+    auth-module.instructions.md    # Module-specific context
+    payments-module.instructions.md
+    search-module.instructions.md
+```
+
+Each `.instructions.md` file can be configured in two ways:
+
+**Auto-applied instructions** use `applyTo` to target specific files or patterns:
+```markdown
+---
+applyTo: "src/auth/**/*"
+description: "Authentication module guidelines"
+---
+# Authentication Module Instructions
+...
+```
+
+**Context-triggered instructions** use only `description` and are automatically loaded by the AI when the description matches the context:
+```markdown
+---
+description: "Dependency injection patterns and best practices"
+---
+# Dependency Injection Guidelines
+Shared concepts and patterns used across the entire codebase...
+```
+
+Context-triggered instructions are ideal for cross-cutting concerns like dependency injection, logging patterns, error handling strategies, or architectural concepts. The AI agent can load these automatically when it determines the description is relevant to the current task or conversation context.
+
+<!-- TODO: Add step-by-step guide for organizing existing documentation into instruction files -->
+
+### Context inheritance patterns
+
+VS Code combines all applicable instruction files automatically:
```

Comment:
Is it concatenated in this order? That was a question I got from developers, and I ended up going into Chat Debug View to try to answer their question. If we can be clear about the order, developers will appreciate it.

Reply:
Its not, there is no priority order or callout in the instructions; and each model has slightly different best practices (like mention important things on top of the system prompt and again in the end).

Once make it clearer as we instruct the model on what takes precedence (like AGENTS.md in root vs in sub-folders vs chat mode, etc); and share that with users.